File: backend/app/api/planner.py
"""app/api/planner.py

FastAPI router for planner endpoints.
Handles planner entry operations (adding/removing meals from planner).
Meal CRUD is handled by the meals router.
"""

from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.api.dependencies import get_current_user
from app.database.db import get_session
from app.dtos.planner_dtos import (
    CookingStreakDTO,
    PlannerBulkAddDTO,
    PlannerEntryResponseDTO,
    PlannerOperationResultDTO,
    PlannerReorderDTO,
    PlannerSummaryDTO,
)
from app.models.user import User
from app.services.planner_service import (
    InvalidMealError,
    PlannerFullError,
    PlannerService,
)

router = APIRouter()


# -- Read Operations -----------------------------------------------------------------------------
@router.get("/entries", response_model=List[PlannerEntryResponseDTO])
def get_all_entries(
    meal_id: Optional[int] = Query(None, description="Filter by meal ID"),
    completed: Optional[bool] = Query(None, description="Filter by completion status"),
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Get planner entries with optional filtering.

    - **meal_id**: Filter entries for a specific meal
    - **completed**: Filter by completion status (true/false)
    """
    service = PlannerService(session, current_user.id)
    return service.get_all_entries(meal_id=meal_id, completed=completed)


@router.get("/entries/{entry_id}", response_model=PlannerEntryResponseDTO)
def get_entry(
    entry_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Get a single planner entry by ID."""
    service = PlannerService(session, current_user.id)
    entry = service.get_entry(entry_id)
    if not entry:
        raise HTTPException(status_code=404, detail="Planner entry not found")
    return entry


@router.get("/summary", response_model=PlannerSummaryDTO)
def get_summary(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Get a summary of the current planner state."""
    service = PlannerService(session, current_user.id)
    return service.get_summary()


@router.get("/meal-ids", response_model=List[int])
def get_meal_ids(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Get all meal IDs currently in the planner."""
    service = PlannerService(session, current_user.id)
    return service.get_meal_ids()


@router.get("/streak", response_model=CookingStreakDTO)
def get_cooking_streak(
    tz: Optional[str] = Query(None, description="User's IANA timezone (e.g., 'America/New_York')"),
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Get cooking streak information.

    Returns the current consecutive day streak, longest streak ever,
    and activity for the current calendar week (Monday-Sunday).

    The tz parameter ensures dates are calculated in the user's timezone.
    """
    service = PlannerService(session, current_user.id)
    return service.get_cooking_streak(user_timezone=tz)


# -- Add Operations ------------------------------------------------------------------------------
@router.post("/entries/{meal_id}", response_model=PlannerEntryResponseDTO, status_code=201)
def add_meal_to_planner(
    meal_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Add a meal to the planner.

    Creates a new planner entry referencing the meal.
    The meal must exist. Maximum 15 entries allowed.
    """
    service = PlannerService(session, current_user.id)
    try:
        return service.add_meal_to_planner(meal_id)
    except PlannerFullError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except InvalidMealError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except RuntimeError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/entries/bulk", response_model=List[PlannerEntryResponseDTO], status_code=201)
def add_meals_to_planner(
    data: PlannerBulkAddDTO,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Add multiple meals to the planner.

    All meals must exist. Total entries cannot exceed 15.
    """
    service = PlannerService(session, current_user.id)
    try:
        return service.add_meals_to_planner(data.meal_ids)
    except PlannerFullError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except InvalidMealError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except RuntimeError as e:
        raise HTTPException(status_code=500, detail=str(e))


# -- Update Operations ---------------------------------------------------------------------------
@router.put("/entries/reorder", response_model=PlannerOperationResultDTO)
def reorder_entries(
    data: PlannerReorderDTO,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Reorder planner entries.

    Provide entry IDs in the desired order. Positions will be reassigned 0, 1, 2, ...
    """
    service = PlannerService(session, current_user.id)
    success = service.reorder_entries(data.entry_ids)
    return PlannerOperationResultDTO(
        success=success,
        message="Entries reordered successfully" if success else "Failed to reorder entries",
        affected_count=len(data.entry_ids) if success else 0,
    )


@router.post("/entries/{entry_id}/complete", response_model=PlannerEntryResponseDTO)
def mark_completed(
    entry_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Mark a planner entry as completed."""
    service = PlannerService(session, current_user.id)
    entry = service.mark_completed(entry_id)
    if not entry:
        raise HTTPException(status_code=404, detail="Planner entry not found")
    return entry


@router.post("/entries/{entry_id}/incomplete", response_model=PlannerEntryResponseDTO)
def mark_incomplete(
    entry_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Mark a planner entry as incomplete."""
    service = PlannerService(session, current_user.id)
    entry = service.mark_incomplete(entry_id)
    if not entry:
        raise HTTPException(status_code=404, detail="Planner entry not found")
    return entry


@router.post("/entries/{entry_id}/cycle-shopping-mode", response_model=PlannerEntryResponseDTO)
def cycle_shopping_mode(
    entry_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Cycle the shopping mode of a planner entry.

    Cycles through: all -> produce_only -> none -> all
    - all: Include all ingredients in shopping list
    - produce_only: Include only produce category ingredients
    - none: Exclude from shopping list entirely
    """
    service = PlannerService(session, current_user.id)
    entry = service.cycle_shopping_mode(entry_id)
    if not entry:
        raise HTTPException(status_code=404, detail="Planner entry not found")
    return entry


# -- Remove Operations ---------------------------------------------------------------------------
@router.delete("/entries/{entry_id}")
def remove_entry(
    entry_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """
    Remove a planner entry.

    Note: This does NOT delete the underlying meal.
    The meal can be re-added to the planner later.
    """
    service = PlannerService(session, current_user.id)
    if not service.remove_entry(entry_id):
        raise HTTPException(status_code=404, detail="Planner entry not found")
    return {"message": "Entry removed from planner"}


@router.delete("/entries/by-meal/{meal_id}")
def remove_entries_by_meal(
    meal_id: int,
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Remove all planner entries for a specific meal."""
    service = PlannerService(session, current_user.id)
    count = service.remove_entries_by_meal(meal_id)
    return {"message": f"Removed {count} entries for meal {meal_id}"}


@router.delete("/clear")
def clear_planner(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Clear all entries from the planner."""
    service = PlannerService(session, current_user.id)
    count = service.clear_planner()
    return {"message": f"Cleared {count} entries from planner"}


@router.delete("/clear-completed")
def clear_completed(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_user),
):
    """Clear all completed entries from the planner."""
    service = PlannerService(session, current_user.id)
    count = service.clear_completed()
    return {"message": f"Cleared {count} completed entries from planner"}

File: frontend/src/lib/api/planner.ts
import type {
  MealSelectionResponseDTO,
  MealSelectionCreateDTO,
  MealSelectionUpdateDTO,
  MealPlanSummaryDTO,
} from "@/types/meal";
import type { PlannerEntryResponseDTO } from "@/types/planner";
import type { CookingStreakDTO } from "@/types/recipe";
import { fetchApi } from "./base";

export const plannerApi = {
  /**
   * Get all meals, optionally filtered by saved status
   * @param filters - Optional filter parameters
   * @param token - Optional auth token for authenticated requests
   */
  getMeals: (filters?: { saved?: boolean }, token?: string | null): Promise<MealSelectionResponseDTO[]> => {
    const query = filters?.saved !== undefined ? `?saved=${filters.saved}` : "";
    return fetchApi<MealSelectionResponseDTO[]>(`/api/meals${query}`, undefined, token);
  },

  /**
   * Get planner summary
   * @param token - Optional auth token for authenticated requests
   */
  getSummary: (token?: string | null): Promise<MealPlanSummaryDTO> =>
    fetchApi<MealPlanSummaryDTO>("/api/planner/summary", undefined, token),

  /**
   * Get a single meal
   * @param id - Meal ID
   * @param token - Optional auth token for authenticated requests
   */
  getMeal: (id: number, token?: string | null): Promise<MealSelectionResponseDTO> =>
    fetchApi<MealSelectionResponseDTO>(`/api/meals/${id}`, undefined, token),

  /**
   * Create a new meal
   * @param data - Meal data
   * @param token - Optional auth token for authenticated requests
   */
  createMeal: (data: MealSelectionCreateDTO, token?: string | null): Promise<MealSelectionResponseDTO> =>
    fetchApi<MealSelectionResponseDTO>(
      "/api/meals",
      {
        method: "POST",
        body: JSON.stringify(data),
      },
      token
    ),

  /**
   * Update an existing meal
   * @param id - Meal ID
   * @param data - Meal data to update
   * @param token - Optional auth token for authenticated requests
   */
  updateMeal: (
    id: number,
    data: MealSelectionUpdateDTO,
    token?: string | null
  ): Promise<MealSelectionResponseDTO> =>
    fetchApi<MealSelectionResponseDTO>(
      `/api/meals/${id}`,
      {
        method: "PUT",
        body: JSON.stringify(data),
      },
      token
    ),

  /**
   * Delete a meal
   * @param id - Meal ID
   * @param token - Optional auth token for authenticated requests
   */
  deleteMeal: (id: number, token?: string | null): Promise<void> =>
    fetchApi<void>(`/api/meals/${id}`, { method: "DELETE" }, token),

  /**
   * Toggle meal saved status (saved meals persist after leaving planner)
   * @param id - Meal ID
   * @param token - Optional auth token for authenticated requests
   */
  toggleSave: (id: number, token?: string | null): Promise<MealSelectionResponseDTO> =>
    fetchApi<MealSelectionResponseDTO>(
      `/api/meals/${id}/save`,
      {
        method: "POST",
      },
      token
    ),

  /**
   * Add a recipe as a side dish to a meal
   * @param mealId - Meal ID
   * @param recipeId - Recipe ID to add as side
   * @param token - Optional auth token for authenticated requests
   */
  addSideToMeal: (mealId: number, recipeId: number, token?: string | null): Promise<MealSelectionResponseDTO> =>
    fetchApi<MealSelectionResponseDTO>(
      `/api/meals/${mealId}/sides/${recipeId}`,
      {
        method: "POST",
      },
      token
    ),

  /**
   * Clear planner entries
   * @param token - Optional auth token for authenticated requests
   */
  clearPlan: (token?: string | null): Promise<void> =>
    fetchApi<void>("/api/planner/clear", { method: "DELETE" }, token),

  /**
   * Remove a meal from the planner by meal ID (meal persists in Meals table)
   * @param mealId - Meal ID
   * @param token - Optional auth token for authenticated requests
   */
  removeFromPlanner: (mealId: number, token?: string | null): Promise<{ message: string }> =>
    fetchApi<{ message: string }>(
      `/api/planner/entries/by-meal/${mealId}`,
      {
        method: "DELETE",
      },
      token
    ),

  // -------------------------------------------------------------------------
  // Planner Entry Methods (Weekly Menu)
  // -------------------------------------------------------------------------

  /**
   * Get all planner entries with hydrated meal data
   * @param token - Optional auth token for authenticated requests
   */
  getEntries: (token?: string | null): Promise<PlannerEntryResponseDTO[]> =>
    fetchApi<PlannerEntryResponseDTO[]>("/api/planner/entries", undefined, token),

  /**
   * Add a meal to the planner (creates a PlannerEntry)
   * @param mealId - Meal ID
   * @param token - Optional auth token for authenticated requests
   */
  addToPlanner: (mealId: number, token?: string | null): Promise<PlannerEntryResponseDTO> =>
    fetchApi<PlannerEntryResponseDTO>(
      `/api/planner/entries/${mealId}`,
      {
        method: "POST",
      },
      token
    ),

  /**
   * Remove a planner entry by entry ID (meal persists)
   * @param entryId - Planner entry ID
   * @param token - Optional auth token for authenticated requests
   */
  removeEntry: (entryId: number, token?: string | null): Promise<{ message: string }> =>
    fetchApi<{ message: string }>(
      `/api/planner/entries/${entryId}`,
      {
        method: "DELETE",
      },
      token
    ),

  /**
   * Mark a planner entry as complete (records recipe history)
   * @param entryId - Planner entry ID
   * @param token - Optional auth token for authenticated requests
   */
  markComplete: (entryId: number, token?: string | null): Promise<PlannerEntryResponseDTO> =>
    fetchApi<PlannerEntryResponseDTO>(
      `/api/planner/entries/${entryId}/complete`,
      {
        method: "POST",
      },
      token
    ),

  /**
   * Mark a planner entry as incomplete
   * @param entryId - Planner entry ID
   * @param token - Optional auth token for authenticated requests
   */
  markIncomplete: (entryId: number, token?: string | null): Promise<PlannerEntryResponseDTO> =>
    fetchApi<PlannerEntryResponseDTO>(
      `/api/planner/entries/${entryId}/incomplete`,
      {
        method: "POST",
      },
      token
    ),

  /**
   * Cycle the shopping mode of a planner entry: all -> produce_only -> none -> all
   * @param entryId - Planner entry ID
   * @param token - Optional auth token for authenticated requests
   */
  cycleShoppingMode: (entryId: number, token?: string | null): Promise<PlannerEntryResponseDTO> =>
    fetchApi<PlannerEntryResponseDTO>(
      `/api/planner/entries/${entryId}/cycle-shopping-mode`,
      {
        method: "POST",
      },
      token
    ),

  /**
   * Clear all completed planner entries
   * @param token - Optional auth token for authenticated requests
   */
  clearCompleted: (token?: string | null): Promise<{ message: string }> =>
    fetchApi<{ message: string }>(
      "/api/planner/clear-completed",
      {
        method: "DELETE",
      },
      token
    ),

  /**
   * Reorder planner entries by providing entry IDs in desired order
   * @param entryIds - Array of entry IDs in desired order
   * @param token - Optional auth token for authenticated requests
   */
  reorderEntries: (
    entryIds: number[],
    token?: string | null
  ): Promise<{ success: boolean; message: string }> =>
    fetchApi<{ success: boolean; message: string }>(
      "/api/planner/entries/reorder",
      {
        method: "PUT",
        body: JSON.stringify({ entry_ids: entryIds }),
      },
      token
    ),

  /**
   * Get cooking streak information
   * Sends user's timezone to ensure correct date calculations
   * @param token - Optional auth token for authenticated requests
   */
  getStreak: (token?: string | null): Promise<CookingStreakDTO> => {
    const tz = Intl.DateTimeFormat().resolvedOptions().timeZone;
    return fetchApi<CookingStreakDTO>(
      `/api/planner/streak?tz=${encodeURIComponent(tz)}`,
      undefined,
      token
    );
  },
};
